beacon-chain/p2p/connection_gater.go

```diff
@@ -81,31 +81,98 @@ func (s *Service) validateDial(addr multiaddr.Multiaddr) bool {
 	return true
 }
 
+var privateCIDRList = []string{
+	// Private ip addresses specified by rfc-1918.
+	// See: https://tools.ietf.org/html/rfc1918
+	"10.0.0.0/8",
+	"172.16.0.0/12",
+	"192.168.0.0/16",
+	// Reserved address space for CGN devices, specified by rfc-6598
+	// See: https://tools.ietf.org/html/rfc6598
+	"100.64.0.0/10",
+	// IPv4 Link-Local addresses, specified by rfc-3926
+	// See: https://tools.ietf.org/html/rfc3927
+	"169.254.0.0/16",
+}
+
 // configureFilter looks at the provided allow lists and
 // deny lists to appropriately create a filter.
 func configureFilter(cfg *Config) (*multiaddr.Filters, error) {
 	addrFilter := multiaddr.NewFilters()
-	// Configure from provided allow list in the config.
-	if cfg.AllowListCIDR != "" {
+	switch {
+	case cfg.AllowListCIDR == "public":
+		cfg.DenyListCIDR = append(cfg.DenyListCIDR, privateCIDRList...)
+	case cfg.AllowListCIDR == "private":
+		err := privateCIDRFilter(addrFilter, multiaddr.ActionAccept)
+		if err != nil {
+			return nil, err
+		}
+	case cfg.AllowListCIDR != "":
 		_, ipnet, err := net.ParseCIDR(cfg.AllowListCIDR)
 		if err != nil {
 			return nil, err
 		}
 		addrFilter.AddFilter(*ipnet, multiaddr.ActionAccept)
 	}
+
 	// Configure from provided deny list in the config.
 	if len(cfg.DenyListCIDR) > 0 {
 		for _, cidr := range cfg.DenyListCIDR {
+			// If an entry in the deny list is "private", we iterate through the
+			// private addresses and add them to the filter. Likewise, if the deny
+			// list is "public", then we add all private address to the accept filter,
+			switch {
+			case cidr == "private":
+				err := privateCIDRFilter(addrFilter, multiaddr.ActionDeny)
+				if err != nil {
+					return nil, err
+				}
+				continue
+			case cidr == "public":
+				err := privateCIDRFilter(addrFilter, multiaddr.ActionAccept)
+				if err != nil {
+					return nil, err
+				}
+				continue
+			}
 			_, ipnet, err := net.ParseCIDR(cidr)
 			if err != nil {
 				return nil, err
 			}
+			action, _ := addrFilter.ActionForFilter(*ipnet)
```

Comment:
need to handle the boolean correctly here again.

Reply:
Since boolean returns false only if the there's no action assigned for the address, in this case I don't think we need to handle the boolean? The purpose of this check is to see if this address had already been assigned an action or not. If it has, and we have a conflict, we log a warning. But no matter what the boolean returns, we still want to assign ActionDeny to the ipnet in the end.

Reply:
Same applies here as answered above.